core/lib/multivm/src/versions/shadow/mod.rs

```diff
@@ -0,0 +1,559 @@
+use std::{
+    cell::RefCell,
+    collections::{BTreeMap, BTreeSet},
+    fmt,
+    sync::Arc,
+};
+
+use zksync_types::{StorageKey, StorageLog, StorageLogWithPreviousValue, Transaction, U256};
+
+use crate::{
+    dump::{VmDump, VmDumpHandler, VmStorageDump},
+    interface::{
+        storage::{ImmutableStorageView, ReadStorage, StoragePtr, StorageView},
+        BootloaderMemory, BytecodeCompressionError, CompressedBytecodeInfo, CurrentExecutionState,
+        FinishedL1Batch, L1BatchEnv, L2BlockEnv, SystemEnv, VmExecutionMode,
+        VmExecutionResultAndLogs, VmFactory, VmInterface, VmInterfaceHistoryEnabled,
+        VmMemoryMetrics,
+    },
+    vm_fast, vm_latest,
+    vm_latest::HistoryEnabled,
+    HistoryMode,
+};
+
+#[cfg(test)]
+mod tests;
+
+struct VmWithReporting<S, Shadow> {
+    vm: Shadow,
+    main_vm_storage: StoragePtr<StorageView<S>>,
+    partial_dump: VmDump,
+    dump_handler: VmDumpHandler,
+    panic_on_divergence: bool,
+}
+
+impl<S: fmt::Debug, Shadow: fmt::Debug> fmt::Debug for VmWithReporting<S, Shadow> {
+    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+        formatter
+            .debug_struct("VmWithReporting")
+            .field("vm", &self.vm)
+            .field("panic_on_divergence", &self.panic_on_divergence)
+            .finish_non_exhaustive()
+    }
+}
+
+impl<S: ReadStorage, Shadow: VmInterface> VmWithReporting<S, Shadow> {
+    fn finalize_dump(&mut self, used_contract_hashes: Vec<U256>) {
+        self.partial_dump.set_storage(VmStorageDump::new(
+            &self.main_vm_storage,
+            used_contract_hashes,
+        ));
+    }
+
+    fn report(mut self, used_contract_hashes: Vec<U256>, err: anyhow::Error) {
+        let batch_number = self.partial_dump.l1_batch_number();
+        tracing::error!("VM execution diverged on batch #{batch_number}!");
+        self.finalize_dump(used_contract_hashes);
+        (self.dump_handler)(self.partial_dump);
+
+        if self.panic_on_divergence {
+            panic!("{err:?}");
+        } else {
+            tracing::error!("{err:#}");
+            tracing::warn!(
+                "New VM is dropped; following VM actions will be executed only on the main VM"
+            );
```

Comment:
If you want, you can use `zksync_vlog::sentry::capture_message` to emit a sentry event to make it easier to notice.